### PR TITLE
Coldcard support

### PR DESCRIPTION
**Pull request by humanumbrella**
## Does this PR introduce a breaking change?

* [ ] Yes
* [ ] No

## Does this PR fix an open issue?

* [ ] Yes
* [ ] No

### CODE REVIEW

File: src/components/Wallet/DirectDeviceExtendedPublicKeyImporter.jsx [hunk 1]
```diff
@@ -7,19 +7,16 @@ import {
   ERROR,
```
Comment (dhruvbansal) [inline src/components/Wallet/DirectDeviceExtendedPublicKeyImporter.jsx]: Try `DirectExtendedPublicKeyImporter`

File: src/components/Wallet/IndirectDeviceExtendedPublicKeyImporter.jsx [hunk 2]
```diff
@@ -0,0 +1,263 @@
+import React from "react";
```
Comment (dhruvbansal) [inline src/components/Wallet/IndirectDeviceExtendedPublicKeyImporter.jsx]: We need to make this component abstract and not concrete.  Add a required prop `Reader` to the component that concrete classes (e.g. `HermitExtendedPublicKeyImporter` or `ColdcardExtendedPublicKeyImporter`) can use to pass in a custom component to handle request/parse cycle.

```
// in IndirectExtendedPublicKeyImporter.jsx

class IndirectExtendedPublicKeyImporter extends React.Component {

  render = () => {
    const {Reader} = this.props;
    return (
      <div>
        <Reader 
          hasError={this.hasBIP32PathError()} 
          onReceive={this.onReceive} 
          interaction={this.interaction()} />
      </div>
    );
  }

}

// in coldcard.js

class ColdcardExtendedPublicKeyImporter extends React.Component {

  defaultBIP32Path = () => {
    const {network, addressType} = this.props;
    // Coldcard-specific default...
  }

  render = () => {
    return (
      <IndirectExtendedPublicKeyImporter
        defaultBIP32Path={this.defaultBIP32Path()}
        Reader={ColdcardFileReader}
      />
    );
  }

}


// in hermit.js

class HermitExtendedPublicKeyImporter extends React.Component {

  render = () => {
    return (
      <IndirectExtendedPublicKeyImporter
        Reader={HermitReader}
      />
    );
  }

}
```

File: src/components/ScriptExplorer/DirectHardwareWalletSignatureImporter.jsx [hunk 3]
```diff
@@ -1,6 +1,5 @@
 import React from "react";
 import PropTypes from "prop-types";
-
```
Comment (dhruvbansal) [inline src/components/ScriptExplorer/DirectHardwareWalletSignatureImporter.jsx]: This file/component should be `DirectSignatureImporter`

File: src/components/ScriptExplorer/IndirectHardwareWalletSignatureImporter.jsx [hunk 4]
```diff
@@ -0,0 +1,353 @@
+import React from "react";
```
Comment (dhruvbansal) [inline src/components/ScriptExplorer/IndirectHardwareWalletSignatureImporter.jsx]: Should be named `IndirectSignatureImporter`

File: src/actions/extendedPublicKeyImporterActions.js [hunk 5]
```diff
@@ -37,3 +42,23 @@ export function setExtendedPublicKeyImporterVisible(value) {
     value,
   };
 }
+
+export function generateRichExtendedPublicKeys(extendedPublicKeyImporters) {
```
Comment (dhruvbansal) [inline src/actions/extendedPublicKeyImporterActions.js]: Move this to the `WalletGenerator` component.

File: src/components/TestSuiteRun/Seed.jsx [hunk 6]
```diff
@@ -17,26 +17,34 @@ class SeedBase extends React.Component {
...
+              {bip39Phrase.slice(12, 18).map(SeedBase.renderSeedWord)}
+            </ol>
+          </Grid>
+          <Grid item md={3}>
+            <ol start={19}>
+              {bip39Phrase.slice(18, 24).map(SeedBase.renderSeedWord)}
+            </ol>
+          </Grid>
       </Grid>
+      <Grid style={{marginTop:'2em', marginBottom:'2em'}}>
```
Comment (dhruvbansal) [inline src/components/TestSuiteRun/Seed.jsx]: Put a `if coldcard` guard on this

**Review by dhruvbansal — CHANGES_REQUESTED** [review]

File: src/components/Wallet/IndirectDeviceExtendedPublicKeyImporter.jsx (hunk 2 above)
Comment (humanumbrella) [reply src/components/Wallet/IndirectDeviceExtendedPublicKeyImporter.jsx]: will review what I did re: this more tomorrow, but I have something working for now.

File: src/components/Hermit/HermitExtendedPublicKeyImporter.jsx [hunk 7]
```diff
@@ -1,8 +1,6 @@
 import React from "react";
```
Comment (dhruvbansal) [inline src/components/Hermit/HermitExtendedPublicKeyImporter.jsx:1]: Add a note why you're not using the shared indirect helper component class (no need to input BIP32 path!)

File: src/components/ScriptExplorer/SignatureImporter.jsx [hunk 8]
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
       errback("Not enough signatures (must be exactly one for each input).");
```
Comment (dhruvbansal) [inline src/components/ScriptExplorer/SignatureImporter.jsx]: "Must be at least one signature for each input"

File: src/components/ScriptExplorer/SignatureImporter.jsx [hunk 9]
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
...
       return;
     }
-    if (inputsSignatures.length > inputs.length) {
-      errback("Too many signatures (must be exactly one for each input).");
-      return;
+
+    if (inputsSignatures.length % inputs.length !== 0) {
+      errback(
+        "Not the right number of signatures (must be exactly one for each input)."
```
Comment (dhruvbansal) [inline src/components/ScriptExplorer/SignatureImporter.jsx]: "Number of signatures must be a multiple of number of inputs."

Comment (dhruvbansal) [reply src/components/ScriptExplorer/SignatureImporter.jsx]: Also add a `return` to short-circuit out of this function after calling `errback`

File: src/components/ScriptExplorer/SignatureImporter.jsx [hunk 10]
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
...
+      );
     }
 
+    const numSignatureSets = inputsSignatures.length / inputs.length;
```
Comment (dhruvbansal) [inline src/components/ScriptExplorer/SignatureImporter.jsx:330]: `numSignatureSets` cannot be larger than number of required signers (also == to the number of signature importers)

File: src/components/ScriptExplorer/SignatureImporter.jsx [hunk 11]
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
...
+      setComplete(number, {
+        signature: inputsSignatures,
+        publicKeys,
+        finalized: true,
+      });
+    } else {
+      // We land here if a PSBT has been uploaded with multiple signature sets.
+      // In case we already have some signatures, filter those out.
+      let signaturesToCheck = this.filterKnownSignatures(inputsSignatures);
+      let sigNumber = number; // so we can iterate, number is a const from props.
```
Comment (dhruvbansal) [inline src/components/ScriptExplorer/SignatureImporter.jsx]: This is `signatureSetNumber`

**Review by dhruvbansal — CHANGES_REQUESTED** [review]

**Pull request closed without merging**